### PR TITLE
tagr.list() functionality

### PR DESCRIPTION
**Pull request by ericlee0112**
# Description
This is part 1 of 2 for https://github.com/tagr-dev/tagr/issues/10
Added .list() method for tagr so now users can get a list of objects by looking up a specific directory in local storage or aws s3. After a user flushes metadata for a particular experiment, the user can then lookup what metadata they have pushed in previous experiments, allowing for granularity up to the tag level in a folder directory path. ```{project}/{experiment}/{tag}```

Here's an example:
Lets say I flushed some metadata to s3:
```
tag.flush('waterflow-tagr', 'eric', 'aws', 'second_experiment_tag')
# project : waterflow-tagr
# experiment : eric
# dump : aws
# tag : second_experiment_tag
```
Now, I want to see what metadata I have pushed for ```waterflow-tagr/eric/second_experiment_tag```
```
tag.list('waterflow-tagr', 'eric', 'second_experiment_tag', 'aws')
```
which returns:
```
['eric/second_experiment_tag/X_train.csv',
 'eric/second_experiment_tag/df_summary.json',
 'eric/second_experiment_tag/linmodel.pkl',
 'eric/second_experiment_tag/model.pkl',
 'eric/second_experiment_tag/y_pred.csv',
 'eric/second_experiment_tag/y_train.csv']
```

# What's changed
- Added .list() methods for both ```aws.py``` and ```local.py``` classes
- In the case of aws, we needed to figure out a way to lookup s3 object paths programmatically. I used the helper functions provided in here https://mageswaran1989.medium.com/list-s3-folders-with-boto3-b2d7fbd631e9 as a template to help us lookup s3 object paths. Made some simplifications for our use case, but its still a lot of code. Maybe we can figure out an even simpler way that requires less code in the future?
- Edited a few docstrings

# Some remarks
Yeah I know, we should probably refactor .flush() so that the ordering of the input parameters should be the same for both flush() and list(). Right now it looks like this:
```
def flush(self, proj, experiment, dump='local', tag=None):
def list(self, proj, experiment, tag="", dump='local'):
```
 We can do that in a future PR

### CODE REVIEW

File: tagr/storage/aws.py [hunk 1]
```diff
@@ -5,11 +5,11 @@
 import json
 from io import StringIO
 import boto3
-
+from tagr.storage.aws_helper import AwsHelper
```
Comment (SunriseLong) [inline tagr/storage/aws.py:8]: add line bread

File: tagr/storage/aws.py [hunk 2]
```diff
@@ -63,3 +63,19 @@ def dump_pickle(self, pickle_object, proj, experiment, tag, filename):
         self.S3.Object(proj, "{}/{}/{}.pkl".format(experiment, tag, filename)).put(
             Body=pickle_object
         )
+
+    def list(self, proj, experiment, tag):
+        '''
```
Comment (SunriseLong) [inline tagr/storage/aws.py:68]: => double quotes

File: tagr/storage/aws_helper.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
+'''
```
Comment (SunriseLong) [inline tagr/storage/aws_helper.py:1]: => double quotes

File: tagr/storage/aws_helper.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+code heavy helper functions for Aws.py
+
```
Comment (SunriseLong) [inline tagr/storage/aws_helper.py:3]: remove blank line

File: tagr/storage/local.py [hunk 5]
```diff
@@ -50,6 +50,23 @@ def dump_pickle(self, pickle_object, proj, experiment, tag, filename):
             "{}/{}/{}/{}.pkl".format(proj, experiment, tag, filename), "wb"
         ))
     
+    def list(self, proj, experiment, tag):
+        '''
```
Comment (SunriseLong) [inline tagr/storage/local.py:54]: => double quotes

File: tagr/tagging/artifacts.py [hunk 6]
```diff
@@ -166,3 +166,25 @@ def flush(self, proj, experiment, dump='local', tag=None):
...
+            - for dump, asssume local by default if destination not provided
+        """
+        # determine which storage provider to use
+        if dump == 'aws':
+            self.storage_provider = Aws()
+        elif dump == 'local':
+            self.storage_provider = Local()
+            
+        return self.storage_provider.list(proj, experiment, tag)
+        #return self.storage_provider.fetch(proj, path)
```
Comment (SunriseLong) [inline tagr/tagging/artifacts.py:189]: remove

File: tagr/storage/aws.py [hunk 7]
```diff
@@ -5,11 +5,11 @@
 import json
 from io import StringIO
 import boto3
-
+from tagr.storage.aws_helper import AwsHelper
 from tagr.utils import NpEncoder
 
 csv_buffer = StringIO()
-
+aws_helper = AwsHelper()
```
Comment (SunriseLong) [inline tagr/storage/aws.py:12]: instantiate within list() method

File: tagr/storage/aws_helper.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        if isinstance(object_path, str):
+            kwargs['Prefix'] = object_path
+        else:
+            kwargs['Prefix'] = str(object_path)
+
+        kwargs['MaxKeys'] = max_keys_per_request
+
+        while True:
+
+            # The S3 API response is a large blob of metadata.
```
Comment (SunriseLong) [inline tagr/storage/aws_helper.py:33]: blob or a dict of objects?

File: tagr/storage/aws.py [hunk 9]
```diff
@@ -63,3 +63,19 @@ def dump_pickle(self, pickle_object, proj, experiment, tag, filename):
         self.S3.Object(proj, "{}/{}/{}.pkl".format(experiment, tag, filename)).put(
             Body=pickle_object
         )
+
+    def list(self, proj, experiment, tag):
```
Comment (SunriseLong) [inline tagr/storage/aws.py:67]: I see we have a method named `list()` in `artifacts.py`. lets make this naming explicit. rename to private function

File: tagr/storage/local.py [hunk 10]
```diff
@@ -50,6 +50,23 @@ def dump_pickle(self, pickle_object, proj, experiment, tag, filename):
             "{}/{}/{}/{}.pkl".format(proj, experiment, tag, filename), "wb"
         ))
     
+    def list(self, proj, experiment, tag):
```
Comment (SunriseLong) [inline tagr/storage/local.py:53]: I see we have a method named `list()` in `artifacts.py`. lets make this naming explicit. rename to private function

**Review by SunriseLong — COMMENTED** [review]
Good work on the the list method. That s3 skd stuff is real annoying

File: tagr/storage/aws_helper.py (hunk 8 above)
Comment (ericlee0112) [reply tagr/storage/aws_helper.py:33]: After debugging, found out its a dict, idk this comment was just a copypasta

**Review by SunriseLong — APPROVED** [review]
Great work!

**Pull request merged**